src/analyses/variable-sensitivity/abstract_environment.cpp

```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+                               {ID_le, ID_gt},
+                               {ID_lt, ID_ge},
+                               {ID_ge, ID_lt},
+                               {ID_gt, ID_le}};
+
+exprt invert_expr(const exprt &expr)
+{
+  auto expr_id = expr.id();
+  auto inverse_operation = inverse_operations.find(expr_id);
+  if(inverse_operation == inverse_operations.end())
+    return nil_exprt();
+
+  auto relation_expr = to_binary_relation_expr(expr);
+  auto inverse_op = inverse_operation->second;
+  return binary_relation_exprt(
+    relation_expr.lhs(), inverse_op, relation_expr.rhs());
+}
+
+exprt assume_not(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto not_expr = to_not_expr(expr);
+
+  auto inverse_expression = invert_expr(not_expr.op());
+  if(inverse_expression.is_not_nil())
+    return env.do_assume(inverse_expression, ns);
+
+  auto result = env.do_assume(not_expr.op(), ns);
+  if(result.is_boolean())
+    result = result.is_true() ? exprt(false_exprt()) : true_exprt();
+  return result;
+}
+
+exprt assume_and(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto and_expr = to_and_expr(expr);
+  for(auto const &operand : and_expr.operands())
+  {
+    auto result = env.do_assume(operand, ns);
+    if(result.is_nil() || result.is_false())
+      return result;
+  }
+  return true_exprt();
+}
+
+exprt assume_or(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  exprt or_result = false_exprt();
+  auto or_expr = to_or_expr(expr);
+  for(auto const &operand : or_expr.operands())
+  {
+    auto result = env.do_assume(operand, ns);
+    if(result.is_nil())
+      return result;
+    if(result.is_true())
+      or_result = true_exprt();
+  }
+  return or_result;
+}
+
+exprt assume_eq(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto const &equal_expr = to_binary_expr(expr);
+
+  auto left = env.eval(equal_expr.lhs(), ns);
+  auto right = env.eval(equal_expr.rhs(), ns);
+
+  if(left->is_top() || right->is_top())
+    return nil_exprt();
+  if(!is_value(left) || !is_value(right))
+    return nil_exprt();
+
+  auto meet = left->meet(right);
+
+  if(meet->is_bottom())
+    return false_exprt();
+
+  if(is_lvalue(equal_expr.lhs()))
+    env.assign(equal_expr.lhs(), meet, ns);
+  if(is_lvalue(equal_expr.rhs()))
+    env.assign(equal_expr.rhs(), meet, ns);
+  return true_exprt();
+}
+
+exprt assume_noteq(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto const &notequal_expr = to_binary_expr(expr);
+
+  auto left = env.eval(notequal_expr.lhs(), ns);
+  auto right = env.eval(notequal_expr.rhs(), ns);
+
+  if(left->is_top() || right->is_top())
+    return nil_exprt();
+  if(!is_value(left) || !is_value(right))
+    return nil_exprt();
+
+  auto meet = left->meet(right);
+
+  if(meet->is_bottom())
+    return true_exprt();
+
+  return false_exprt();
```

Comment:
Is this correct?  If left is `{0, 1, 2}` and right is `{1, 2, 3}`, the meet is `{1,2}` which is not enough to conclude anything.  We can only return `false_exprt()` if both left and right are the same, single possible value.

Reply:
I believe it is, but that doesn't mean it is.
For the equals case, if objects meet we consider that there exists some equality and so we return true. If the meet is empty, they're not equal and we return false. 
From our conversations, I understood the != case to be essentially equivalent. If the object do not meet, they are not equal and we return true. If there's some meet they are not not equal and we return false. 

Reworking this to return `false_exprt()` only when left and right are the same and `nil_exprt` otherwise breaks (or at least changes) existing behaviour in, for example, some of the simplification regression tests. It also causes segfaults in some of the regression tests I wrote around this work. 

The code, as written here, behaves better than the previous code in many cases and no worse than the previous code in every other case. It may not be absolutely correct, but I feel we can still merge this change, gain the benefits, and then revisit as work continues.

Reply:
There is an asymmetry which happens because the domains are representing /possible/ values for each variable rather than values they definitely can take.

So in the equal cases, we can take the meet of the two domains and get all of the values at which they could both have and be equal.  As you say, if it is empty then there are no ways they can be equal, so we return false and set the domain to bottom.  If there are things in the meet then it is possible they are equal and /we know that they must have a value from the meet/, so we can prune as you do.


The inequality case is a little bit harder and less useful.  If we take the meet of the values and it is empty then there was never any chance they were equal, so we return true because that constraint really is always true.  That much is symmetric.  However if the meet is non-empty (not bottom) then all is says is that it is /possible/ that they could be equal.  It doesn't mean they will be.  In the example I give, if the meet is `{1,2}` then it is possible that `left` is `1` and `right` is `2` or vica versa.  The only cases we need to remove are when they /have/ to be equal.  However that is really hard to represent if you have a set of alternatives for `left` and a separate one for `right` (this is where the non-relational part of the abstraction is an impediment, we don't have a way of representing the relationship between them).  To prune out possibilities we need to check when things are forced to be equal.  This only happens when one side only has one element (i.e. `to_constant` returns something other than `ID_nil`).  If the other side is also one element and the same one (meet is not empty) then they are necessarily equal, so we can return false as assuming they are not equal rules out all possibilities.  ( There is also a pruning case, where one side has one possibility and the other has multiple ones, in which case you can remove that value.  Easy-ish for value-set, possible for intervals if it is the boundary, otherwise not really possible. )

Whether this gets fixed as part of this PR or as a follow up ... either is fine.  I have a mild preference for fixing it now, esp. given the other changes I am requesting to this area.